### DISCUSSION TITLE
Architectural question

### DISCUSSION THREAD

**Discussion by johntraynor** [Kitware/trame#288] [General]
Hi all,

First just want to say this looks like a great framework

I have been evaluating whether I can use trame to build an application for 3D vision scanning and analysis of small parts. I would be very interesting in hearing peoples views on whether it is a good fit.

So a little about the application

- The application will have to communicate with third part 3D scanners with .netcore app running in the back ground on a local pc
- The application front end will run in a browser and connects to a .netcore back end server by means of a websocket /  SignalR / similar
- UI needs to have multiple page views ( I presume I can use [trame-router](https://github.com/Kitware/trame-router/tree/master/trame)
- I need to be able to load the point cloud scanned and fit planes to parts of the point cloud that the user selects (Can I use vtk to interact and select segments of the point cloud? Can it fit a plane? Does trame have a widget?) 
- Do I need PCL to fit planes  - if so how would I tie it all together in trame?
- With these created planes I need to measure things like step heights etc - can I do this with trame? Are there any widgets available to help simplify? 

I'm familiar with Vue (more experience with angular) and basic understanding of vtk (visualizing models etc...). My biggest challenge I think is creating a visual interface that is easy for the user to interact with the point cloud and extract the planes, and then create tools for measuring heights etc.

So is trame a good fit or am I using it for more than what it was envisaged to do?

Thanks

JT

-

**Comment by jourdain** [comment]
Hi JT,

Thanks for looking into trame and I'll try to provide some feedback on what I know.

So trame could definitely work with what you are looking for, but it won't be a silver bullet either. Let me explain what trame is offering and what will be remaining on your side to some extent.

First of all trame really focus on connecting a Vue.js front end to a Python backend. All of that can be done by writing simple Python code and "skip" the web craziness. If frontend pieces are missing, you can always add your own by creating a vue plugin exposing your own widgets/components and enabling it in trame (no custom API to follow). 
Technically, if you love frontend (react, angular, ...), you can build your own and just use the trame server side.

Then on top of that, trame has the best in-class integration with VTK/ParaView by providing components for doing Remote or Local rendering. The local rendering option is still lacking some capabilities but still provide some reasonable support coverage. For instance if you intent to use and interact with 3D widgets, you will be better to use remote rendering.

Now, for you if VTK is a good fit for dealing with your data and interaction (3d widgets), then trame will be really good as you will code your app like a regular desktop app using Python to connect the various pieces. This mean, you can involve other libraries that have a Python interface like PCL.

I'm not sure if I missed anything from your question list but we also have some algo and infrastructure available for point cloud that get exposed via [VeloView](https://www.paraview.org/veloview/) which is based on ParaView. Since I'm not familiar with that part, I won't add more. But I would assume pieces of what you are looking for might be available already and PCL might not be needed after all?
And like before, something like VeloView could be leveraged by trame. 

HTH,

Seb

**Comment by johntraynor** [comment]
Hi Seb,

Many thanks for your detailed reply. Very much appreciated.

Love the idea of writing Python code that can take control of vue and VTK etc, all from the back end

A couple of questions based on your initial reply  

> Technically, if you love frontend (react, angular, ...), you can build your own and just use the trame server side.

1. If I wanted to stick with angular (a lot of works done already on it so might make sense) how would I go about using trame on the server side - is there a resource you could point me towards? It's not obvious to me how I would embed it on the client side with Typescript / JS. 

> if you intent to use and interact with 3D widgets, you will be better to use remote rendering.

2. Is there a link to readily availablelist of 3D widgets with example on their use case. Of Particular interest are one that may help with point cloud selection etc...

> something like VeloView could be leveraged by trame

3. I'll look into Veloview. Looks like it has some of the functionality I am looking for alright

Thanks in advance

JT

**Comment by jourdain** [comment]
> If I wanted to stick with angular (a lot of works done already on it so might make sense) how would I go about using trame on the server side - is there a resource you could point me towards? It's not obvious to me how I would embed it on the client side with Typescript / JS.

Unfortunately we don't have such example/doc as it is not the way we recommend and it would be very hard to cover every use-cases. But technically it is the same thing as when we are doing VTK/ParaView Web using wslink+vtk.js. Trame just offer a specific set of RPC and Pub/Sub that you can rely on within your client side. 

> Is there a link to readily availablelist of 3D widgets with example on their use case. Of Particular interest are one that may help with point cloud selection etc...

- https://examples.vtk.org/site/Cxx/#widgets
- https://examples.vtk.org/site/Python/#widgets

**Reply by johntraynor** [reply]
Many thanks Seb. I'm starting to get a better handle on all the various components and how they interact

So can I ask you one further question - (I'm like Columbo if you are old enough to remember him!)

I came across this video [link ](https://www.youtube.com/watch?v=BZBQXcBvHW0) where PCL functionality was brought into Paraview as filters and it looks like you can dynamically create various filters / steps and visualize it as it happens. If I wanted to recreate this functionality, but in a web browser, what are the steps I need to follow at a high level.

1. Would I use trame? 
2. Can I use trame to build up this level of functionality or is it for simpler application?
3. Are there web widgets for the pipeline interactions in Paraview etc or would I have to customise the UI myself?
4. Would I instead use the ParaView visualizer as a starting block [see this link](https://kitware.github.io/visualizer/)?

Basically I would like as much much functionality as ParaView offers but in a web browser? At a minimum I would like the functionality in the PLC example video above, but in a web browser

Thanks in advance

JT

**Reply by jourdain** [reply]
1. Trame does not care of what you do with it. But if you aim to make a desktop like application in the web and leverage VTK or ParaView, you definitely should. 
2. Definitely
3. You have both options
4. Probably not, unless that is the UI you want and need. But it seems you know the pipeline you want to use. So it is easier to build it on the back and just present the UI you actually need. 

Are you looking for the generality UI of ParaView or just its capability in a tailor UI?

**Reply by johntraynor** [reply]
Hi Seb,

Thanks again for the reply. 

Yep, looking more for the generality UI of ParaView but in a browser. Is that possible with not too too much work?

Basically looking to do the following to start with

1. Extend the current set of PCL filters available
2. Have the ability to add, remove, move up and down the filters in the pipeline browser while being able to see visualize the result for each of the PCL filters like you can in ParaView desktop, but do this in the web browser
3. Have the option of bring in some of the other functionality standard in ParaView desktop, into the web browser (add render views, load more filters  etc..) - basically as much functionality as the standard Paraview desktop but in the browser - not sure if this is easily possible
4. Doing all the above from a web browser as opposed to inside the desktop application

Hope this clarifies

Thanks 

JT

**Reply by jourdain** [reply]
So you want ParaView as-is in the browser. That would be Visualizer except that we are still behind on that one and we currently don't have any active project supporting it right now. But if you have funds, we can team up so we can keep pushing it forward and you can leverage it for your use case.

The general approach make thing trickier and therefore more expensive than going with a tailored solution. 

Either way, going with ParaView as backend, you will get the power and flexibility that you need. But if you can narrow down your user flow and UI interaction, you can probably achieve what you are looking for more simply than going the generic path.

**Reply by johntraynor** [reply]
Hi Seb,

Thanks for the feedback. Might be worth a quick conference call to see if there was a way to work together on this? I'm away until the end of the month but we could set something very early August if that worked for you and the team?. What's the best way to organise the call?

Thanks

JT

**Reply by jourdain** [reply]
The best way is to reach out [here](https://www.kitware.com/contact/). Thanks and looking forward to chat in person.

**Reply by johntraynor** [reply]
Thanks Seb - I'll chat to you upon my return
